File: frontend/src/speech-text.ts
/* H526 — what a reply sounds like: speech, not read-out markdown (browser side).
 *
 * The twin of `agents/core/voice/speech_text.py`. The hub normalises everything it
 * synthesizes, but the HUD also speaks on its own (the browser's speechSynthesis voice)
 * and splits the live token stream into sentences before any of it reaches the hub — so
 * the same rules run here. `test/speech-text-cases.json` pins both implementations to the
 * same outputs; a change to one must land in the other.
 *
 * `SpeechStreamFilter` is the streaming half: a fenced code block or a reasoning block
 * whose markers arrive split across token deltas is never handed to the sentence
 * aggregator, so no line of it is spoken; an unclosed one at the end is dropped.
 */

const THINK_TAGS = 'think|thinking|reasoning';
const THINK_CLOSED = new RegExp(`<(${THINK_TAGS})\\b[^>]*>[\\s\\S]*?</\\1\\s*>`, 'gi');
const THINK_OPEN = new RegExp(`<(?:${THINK_TAGS})\\b[^>]*>[\\s\\S]*$`, 'i');
const THINK_STRAY_CLOSE = new RegExp(`^[\\s\\S]*</(?:${THINK_TAGS})\\s*>`, 'i');
const FENCE = /(```|~~~)[\s\S]*?(?:\1|$)/g;
const AUTOLINK = /<https?:\/\/[^<>\s]+>/g;
const HTML_TAG = /<\/?[A-Za-z][A-Za-z0-9-]*(?:\s[^<>]*)?\/?>/g;
const IMAGE = /!\[([^\]\n]*)\]\([^)\n]*\)/g;
const LINK = /\[([^\]\n]+)\]\([^)\n]*\)/g;
const URL_RE = /https?:\/\/[^\s<>()[\]]+/g;
const URL_TRAIL = /[.,;:!?'"]+$/;

const RULE = /^([-*_])(?:\s*\1){2,}$/;
const TABLE_SEPARATOR = /^\|?(?:\s*:?-{2,}:?\s*\|)+\s*(?::?-{2,}:?)?\s*\|?$/;
const HEADING = /^#{1,6}\s+(.*?)(?:\s+#+)?$/;
const QUOTE = /^(?:>\s?)+/;
const LIST_MARKER = /^(?:[-*+•]|\d{1,3}[.)])\s+/;
const TASK_BOX = /^\[[ xX]\]\s+/;

// Bold, strikethrough and inline-code markers simply go (their words stay); single-
// character emphasis needs the word-boundary guards so "2*3*4" and "my_var" survive.
const PAIRED_MARKERS = /\*\*|__|~~|`/g;
// Python's \w is Unicode-aware; spell it out so "ș_" behaves the same here.
const ITALIC = /(^|[^\p{L}\p{N}_*])\*(\S(?:.*?\S)?)\*(?![\p{L}\p{N}_*])/gu;
const ITALIC_UNDERSCORE = /(^|[^\p{L}\p{N}_])_(\S(?:.*?\S)?)_(?![\p{L}\p{N}_])/gu;

const EMOJI = /[\u{1F000}-\u{1FAFF}\u{2600}-\u{27BF}\u{2B00}-\u{2BFF}\u{2300}-\u{23FF}\u{FE00}-\u{FE0F}\u{200D}\u{20E3}\u{E0020}-\u{E007F}]+/gu;
const SPACE_BEFORE_PUNCT = /\s+([,.;:!?…])/g;
const SPACES = /\s+/g;
const ENDS_A_SENTENCE = /[.!?…:;,]$/;
const SAYABLE = /[\p{L}\p{N}]/u;          // a letter or a digit, in any script

type Words = { and: string; percent: string; to: string; degrees: string };

/** The words for each symbol, per language (anything not Romanian reads English). */
export const WORDS: Record<'en' | 'ro', Words> = {
  en: { and: 'and', percent: 'percent', to: 'to', degrees: 'degrees' },
  ro: { and: 'și', percent: 'la sută', to: 'spre', degrees: 'grade' },
};

/** "ro" or "en": the request's language, else the voice's ("ro-RO-…"), else the default. */
export function speechLang(lang?: unknown, voice?: unknown, fallback = 'en'): 'ro' | 'en' {
  const voiceHint = typeof voice === 'string' && /^[a-z]{2}-[A-Z]{2}-/.test(voice) ? voice : null;
  for (const hint of [lang, voiceHint, fallback]) {
    if (typeof hint === 'string' && hint.trim()) return hint.trim().toLowerCase().startsWith('ro') ? 'ro' : 'en';
  }
  return 'en';
}

function symbols(line: string, w: Words): string {
  return line
    .replace(/\s*&\s*/g, ` ${w.and} `)
    .replace(/\s*%/g, ` ${w.percent}`)
    .replace(/\s*(?:->|→|=>)\s*/g, ` ${w.to} `)
    .replace(/\s*°\s*C\b/g, ` ${w.degrees} Celsius`)
    .replace(/\s*°\s*F\b/g, ` ${w.degrees} Fahrenheit`)
    .replace(/\s*°/g, ` ${w.degrees}`);
}

function inline(line: string): string {
  return line.replace(PAIRED_MARKERS, '').replace(ITALIC, '$1$2').replace(ITALIC_UNDERSCORE, '$1$2');
}

function oneLine(raw: string, w: Words): string {
  let line = raw.trim();
  if (!line || RULE.test(line)) return '';
  let block = false;
  const heading = HEADING.exec(line);
  if (line.startsWith('|')) {
    if (TABLE_SEPARATOR.test(line)) return '';
    line = line.replace(/^\|+|\|+$/g, '').split('|').map((c) => c.trim()).filter(Boolean).join(', ');
    block = true;
  } else if (heading) {
    line = heading[1];
    block = true;
  } else {
    const quoted = line.replace(QUOTE, '');
    const listed = quoted.replace(LIST_MARKER, '').replace(TASK_BOX, '');
    block = quoted !== line || listed !== quoted;
    line = listed;
  }
  line = inline(line);
  line = line.replace(URL_RE, (m) => `link${(URL_TRAIL.exec(m) || [''])[0]}`);
  line = symbols(line, w);
  line = line.replace(EMOJI, ' ');
  line = line.replace(SPACES, ' ').replace(SPACE_BEFORE_PUNCT, '$1').trim();
  if (block && line && !ENDS_A_SENTENCE.test(line)) line += '.';
  return line;
}

/** `text` as it should be heard, or "" when nothing in it is worth saying. */
export function speechText(text: unknown, lang?: unknown): string {
  let raw = String(text ?? '').replace(/\r\n?/g, '\n');
  raw = raw.replace(THINK_CLOSED, ' ').replace(THINK_OPEN, '').replace(THINK_STRAY_CLOSE, '');
  raw = raw.replace(FENCE, '\n').replace(AUTOLINK, 'link').replace(HTML_TAG, ' ');
  raw = raw.replace(IMAGE, '$1').replace(LINK, '$1');
  const w = WORDS[speechLang(lang)];
  const spoken = raw.split('\n').map((l) => oneLine(l, w)).filter(Boolean).join(' ');
  return SAYABLE.test(spoken) ? spoken : '';   // a lone "." is nothing to say
}

const OPENERS = /```|~~~|<(think|thinking|reasoning)\b[^>]*>/i;
const THINK_CLOSERS: Record<string, RegExp> = {
  think: /<\/think\s*>/i,
  thinking: /<\/thinking\s*>/i,
  reasoning: /<\/reasoning\s*>/i,
};
// A tail that may still become an opener once the next delta arrives.
const MAYBE_OPENER = /(`{1,2}|~{1,2}|<[A-Za-z]{0,9}|<(?:think|thinking|reasoning)\b[^>]*)$/i;

/**
 * Drops fenced code and reasoning from a live token stream, however its markers are
 * split across deltas. `push()` returns the text that is safe to speak so far; a tail
 * that might still turn into a marker is held back until the next delta decides.
 */
export class SpeechStreamFilter {
  private pending = '';
  private closer: RegExp | string | null = null;   // what ends the block we are inside

  push(delta: string): string {
    this.pending += String(delta || '');
    let out = '';
    for (;;) {
      if (this.closer !== null) {
        if (typeof this.closer === 'string') {
          const at = this.pending.indexOf(this.closer);
          if (at < 0) { this.pending = this.pending.slice(-(this.closer.length - 1)); return out; }
          this.pending = this.pending.slice(at + this.closer.length);
        } else {
          const m = this.closer.exec(this.pending);
          if (!m) { this.pending = this.pending.slice(-16); return out; }
          this.pending = this.pending.slice(m.index + m[0].length);
        }
        this.closer = null;
        out += '\n';
        continue;
      }
      const m = OPENERS.exec(this.pending);
      if (m) {
        out += this.pending.slice(0, m.index);
        this.closer = m[1] ? THINK_CLOSERS[m[1].toLowerCase()] : m[0];
        this.pending = this.pending.slice(m.index + m[0].length);
        continue;
      }
      const hold = MAYBE_OPENER.exec(this.pending);
      const cut = hold ? hold.index : this.pending.length;
      out += this.pending.slice(0, cut);
      this.pending = this.pending.slice(cut);
      return out;
    }
  }

  /** The end of the stream: a held tail is text; an unclosed block is never spoken. */
  flush(): string {
    const rest = this.closer === null ? this.pending : '';
    this.pending = '';
    this.closer = null;
    return rest;
  }
}

File: agents/core/voice/speech_text.py
"""H526 — what a reply sounds like: speech, not read-out markdown.

The one normaliser every text-to-speech entry point shares (``TTSEngine.speak`` and
``speak_stream``, so the wake-word pipeline, the voice channel and chat voice notes, and
the ``/tts`` and ``/tts/stream`` routes). ``frontend/src/speech-text.ts`` is its twin
for the HUD's browser voice and live sentence stream; both are pinned by the same cases
(``frontend/src/test/speech-text-cases.json``), so a change here must land there too.

What it does, in order:

1. **Reasoning goes.** ``<think>``/``<thinking>``/``<reasoning>`` blocks, an unclosed
   one to the end, and everything before a closing tag that never opened.
2. **Code goes.** Fenced blocks (```` ``` ```` or ``~~~``), an unclosed fence to the end.
3. **Links read as words.** ``[text](url)`` is its text, ``![alt](url)`` its alt, a bare
   or ``<autolinked>`` URL the word "link" — the URL never eats a closing parenthesis or
   the sentence's full stop. HTML tags go.
4. **Blocks read as sentences.** Headings, list items (``-``, ``*``, ``+``, ``•``,
   ``1.``, ``1)``, task boxes), quotes and table rows lose their markers and end with a
   full stop when they have none; a table's cells are read with commas, its separator row
   and a horizontal rule not at all.
5. **Inline markers go**: bold, italic, strikethrough, inline code (its words stay).
6. **Symbols are said** in the reply's language (Romanian or English): ``&``, ``%``,
   ``->``/``→``/``=>``, ``°``/``°C``/``°F``.
7. **Emoji go** — pictographs, flags, keycaps, skin tones, joiners and variation
   selectors — and the spaces they leave are closed up.

Fish ``[emotion]`` tags are left exactly as they are (``TTSEngine`` strips them for every
other backend). Whitespace is folded, and what has no letter or digit left is ``""``
(nothing to say). The result is a fixed point, so a second pass (a
voice note normalised, then spoken by the engine) changes nothing. Pure: no I/O.
"""
from __future__ import annotations

import re

_THINK_TAGS = r"think|thinking|reasoning"
_THINK_CLOSED = re.compile(rf"<({_THINK_TAGS})\b[^>]*>.*?</\1\s*>", re.S | re.I)
_THINK_OPEN = re.compile(rf"<(?:{_THINK_TAGS})\b[^>]*>.*\Z", re.S | re.I)
_THINK_STRAY_CLOSE = re.compile(rf"\A.*</(?:{_THINK_TAGS})\s*>", re.S | re.I)
_FENCE = re.compile(r"(```|~~~).*?(?:\1|\Z)", re.S)
_AUTOLINK = re.compile(r"<https?://[^<>\s]+>")
_HTML_TAG = re.compile(r"</?[A-Za-z][A-Za-z0-9-]*(?:\s[^<>]*)?/?>")
_IMAGE = re.compile(r"!\[([^\]\n]*)\]\([^)\n]*\)")
_LINK = re.compile(r"\[([^\]\n]+)\]\([^)\n]*\)")
_URL = re.compile(r"https?://[^\s<>()\[\]]+")
_URL_TRAIL = re.compile(r"[.,;:!?'\"]+$")

_RULE = re.compile(r"^([-*_])(?:\s*\1){2,}$")
_TABLE_SEPARATOR = re.compile(r"^\|?(?:\s*:?-{2,}:?\s*\|)+\s*(?::?-{2,}:?)?\s*\|?$")
_HEADING = re.compile(r"^#{1,6}\s+(.*?)(?:\s+#+)?$")
_QUOTE = re.compile(r"^(?:>\s?)+")
_LIST_MARKER = re.compile(r"^(?:[-*+•]|\d{1,3}[.)])\s+")
_TASK_BOX = re.compile(r"^\[[ xX]\]\s+")

#: Bold, strikethrough and inline-code markers simply go (their words stay); single-
#: character emphasis needs the word-boundary guards so ``2*3*4`` and ``my_var`` survive.
_PAIRED_MARKERS = re.compile(r"\*\*|__|~~|`")
_ITALIC = re.compile(r"(^|[^\w*])\*(\S(?:.*?\S)?)\*(?![\w*])")
_ITALIC_UNDERSCORE = re.compile(r"(^|[^\w])_(\S(?:.*?\S)?)_(?!\w)")

_EMOJI = re.compile(
    "["
    "\U0001F000-\U0001FAFF"   # pictographs, emoticons, transport, flags, skin tones, …
    "\u2600-\u27BF"           # miscellaneous symbols, dingbats
    "\u2B00-\u2BFF"           # arrows and stars used as emoji
    "\u2300-\u23FF"           # watches, hourglasses, media keys
    "\uFE00-\uFE0F"           # variation selectors
    "\u200D"                  # zero-width joiner
    "\u20E3"                  # keycap
    "\U000E0020-\U000E007F"   # tag sequences (subdivision flags)
    "]+"
)
_SPACE_BEFORE_PUNCT = re.compile(r"\s+([,.;:!?…])")
_SPACES = re.compile(r"\s+")
_ENDS_A_SENTENCE = re.compile(r"[.!?…:;,]$")
_SAYABLE = re.compile(r"[^\W_]")          # a letter or a digit, in any script

#: The words for each symbol, per language (anything not Romanian reads English).
WORDS = {
    "en": {"and": "and", "percent": "percent", "to": "to", "degrees": "degrees"},
    "ro": {"and": "și", "percent": "la sută", "to": "spre", "degrees": "grade"},
}


def speech_lang(lang: object = None, voice: object = None, *, default: str = "en") -> str:
    """``"ro"`` or ``"en"``: the request's language, else the voice's (``ro-RO-…``),
    else *default*."""
    for hint in (lang, voice if isinstance(voice, str) and re.match(r"^[a-z]{2}-[A-Z]{2}-", voice) else None,
                 default):
        if isinstance(hint, str) and hint.strip():
            return "ro" if hint.strip().lower().startswith("ro") else "en"
    return "en"


def _symbols(line: str, words: dict[str, str]) -> str:
    line = re.sub(r"\s*&\s*", f" {words['and']} ", line)
    line = re.sub(r"\s*%", f" {words['percent']}", line)
    line = re.sub(r"\s*(?:->|→|=>)\s*", f" {words['to']} ", line)
    line = re.sub(r"\s*°\s*C\b", f" {words['degrees']} Celsius", line)
    line = re.sub(r"\s*°\s*F\b", f" {words['degrees']} Fahrenheit", line)
    return re.sub(r"\s*°", f" {words['degrees']}", line)


def _inline(line: str) -> str:
    line = _PAIRED_MARKERS.sub("", line)
    line = _ITALIC.sub(r"\1\2", line)
    return _ITALIC_UNDERSCORE.sub(r"\1\2", line)


def _url(match: re.Match) -> str:
    trail = _URL_TRAIL.search(match.group(0))
    return "link" + (trail.group(0) if trail else "")


def _line(raw: str, words: dict[str, str]) -> str:
    line = raw.strip()
    if not line or _RULE.match(line):
        return ""
    block = False
    if line.startswith("|"):
        if _TABLE_SEPARATOR.match(line):
            return ""
        line = ", ".join(cell.strip() for cell in line.strip("|").split("|") if cell.strip())
        block = True
    elif heading := _HEADING.match(line):
        line, block = heading.group(1), True
    else:
        quoted = _QUOTE.sub("", line)
        block = quoted != line
        listed = _TASK_BOX.sub("", _LIST_MARKER.sub("", quoted))
        block = block or listed != quoted
        line = listed
    line = _inline(line)
    line = _URL.sub(_url, line)
    line = _symbols(line, words)
    line = _EMOJI.sub(" ", line)
    line = _SPACE_BEFORE_PUNCT.sub(r"\1", _SPACES.sub(" ", line)).strip()
    if block and line and not _ENDS_A_SENTENCE.search(line):
        line += "."
    return line


def for_speech(text: object, *, lang: object = None) -> str:
    """*text* as it should be heard, or ``""`` when nothing in it is worth saying."""
    raw = str(text or "").replace("\r\n", "\n").replace("\r", "\n")
    raw = _THINK_CLOSED.sub(" ", raw)
    raw = _THINK_OPEN.sub("", raw)
    raw = _THINK_STRAY_CLOSE.sub("", raw)
    raw = _FENCE.sub("\n", raw)
    raw = _AUTOLINK.sub("link", raw)
    raw = _HTML_TAG.sub(" ", raw)
    raw = _IMAGE.sub(r"\1", raw)
    raw = _LINK.sub(r"\1", raw)
    words = WORDS[speech_lang(lang)]
    lines = (_line(line, words) for line in raw.split("\n"))
    spoken = " ".join(line for line in lines if line)
    return spoken if _SAYABLE.search(spoken) else ""     # a lone "." is nothing to say


__all__ = ["WORDS", "for_speech", "speech_lang"]
